compiler/rustc_data_structures/src/tagged_ptr.rs

```diff
@@ -141,6 +142,40 @@ pub unsafe trait Tag: Copy {
     unsafe fn from_usize(tag: usize) -> Self;
 }
 
+/// Returns the number of bits available for use for tags in a pointer to `T`
+/// (this is based on `T`'s alignment).
+pub const fn bits_for<T: ?Sized + Aligned>() -> u32 {
+    crate::aligned::align_of::<T>().as_nonzero().trailing_zeros()
+}
+
+/// Returns the correct [`Tag::BITS`] constant for a set of tag values.
+pub const fn bits_for_tags(mut tags: &[usize]) -> u32 {
+    let mut bits = 0;
+
+    while let &[tag, ref rest @ ..] = tags {
+        tags = rest;
+
+        let b = bits_for_tag(tag);
+        if b > bits {
+            bits = b;
+        }
+    }
+
+    bits
+}
+
+/// Returns `(size_of::<usize>() * 8) - tag.leading_zeros()`
+const fn bits_for_tag(mut tag: usize) -> u32 {
+    let mut bits = 0;
+
+    while tag > 0 {
+        bits += 1;
+        tag >>= 1;
+    }
+
+    bits
+}
```

Comment:
So, if it can be written shorter (as in doc), why using long version?

Reply:
hm, for some reason I thought that `leading_zeros` is not `const`, oops

Reply:
Someone suggested slightly different version that will be even shorter.
```rust
const fn bits_for_tag(tag: usize) -> u32 {
    if tag == 0 {
        0
    } else {
        (usize::BITS - tag.leading_zeros()) as u32
    }
}
```

Reply:
oh, I forgot that `usize::BITS` exists too lmao

Reply:
It's actually just `usize::BITS - tag.leading_zeros()`. 

`0.leading_zeros() == usize::BITS`  and both of them are already `u32`.